front/lib/resources/user_resource.ts

```diff
@@ -0,0 +1,167 @@
+        id: ids,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static async fetchAllByUsername(username: string): Promise<UserResource[]> {
+    const users = await User.findAll({
+      where: {
+        username,
+      },
+    });
+
+    return users.map((user) => new UserResource(User, user.get()));
+  }
+
+  static fetchAllByEmail(email: string): Promise<UserResource[]> {
+    return User.findAll({
+      where: {
+        email,
+      },
+    }).then((users) => users.map((user) => new UserResource(User, user.get())));
+  }
+
+  static async fetchByExternalId(userId: string): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        sId: userId,
+      },
+    });
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByAuth0Sub(sub: string): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        auth0Sub: sub,
+      },
+    });
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByEmail(email: string): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        email,
+      },
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async fetchByProvider(
+    provider: UserProviderType,
+    providerId: string
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      where: {
+        provider,
+        providerId,
+      },
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  static async getWorkspaceFirstAdmin(
+    workspaceId: number
+  ): Promise<UserResource | null> {
+    const user = await User.findOne({
+      include: [
+        {
+          model: MembershipModel,
+          where: {
+            role: "admin",
+            workspaceId,
+          },
+          required: true,
+        },
+      ],
+      order: [["createdAt", "ASC"]],
+    });
+
+    return user ? new UserResource(User, user.get()) : null;
+  }
+
+  async update(blob: Partial<Attributes<User>>): Promise<void> {
+    const [, affectedRows] = await this.model.update(blob, {
+      where: {
+        id: this.id,
+      },
+      returning: true,
+    });
+
+    Object.assign(this, affectedRows[0].get());
+  }
+
+  async delete(): Promise<Result<undefined, Error>> {
+    try {
+      await this.model.destroy({
+        where: {
+          id: this.id,
+        },
+      });
+
+      return new Ok(undefined);
+    } catch (err) {
+      return new Err(err as Error);
+    }
+  }
+
+  toJSON(): UserType {
+    return {
+      sId: this.sId,
+      id: this.id,
+      createdAt: this.createdAt.getTime(),
+      provider: this.provider,
+      username: this.username,
+      email: this.email,
+      firstName: this.firstName,
+      lastName: this.lastName,
+      fullName: this.firstName + (this.lastName ? ` ${this.lastName}` : ""),
```

Comment:
Can we extract this into a getter so we can use it in different places?

Reply:
Done